### ISSUE TITLE
Remove transaction-wide byte costs

### ISSUE THREAD

**Issue by zbuc** [penumbra-zone/penumbra#3940]
As identified by @hdevalence here: https://github.com/penumbra-zone/web/pull/663#issue-2166001644 if we were to remove transaction-wide byte costs and instead have transaction costs be solely the summation of the actions included in the transaction's costs.

While looking over the work necessary to accomplish this, I noticed that we aren't even including any transaction-wide block cost at the moment:

```rust
impl GasCost for Transaction {
    fn gas_cost(&self) -> Gas {
        self.actions().map(GasCost::gas_cost).sum()
    }
}
```

## Steps

- [ ] Determine block cost per action and set in the various `GasCost` impls

**Comment by aubrika** [comment]
Currently, action execution costs are all hard-coded at `10` - the checkbox here makes sense, but what is the proposed method for determining a more detailed/varied block cost per action? Or asked another way: how should per-action costs ideally be differentiated?

**Comment by hdevalence** [comment]
The block space has cost should be the number of bytes in the proto encoding of the action. 

This can be computed on the Action by doing encode_to_vec().len() and on the ActionPlan by doing the same thing (if the action doesn't require building) or by figuring out what the byte size of the action will be once it is constructed (without actually doing so)

**Comment by hdevalence** [comment]
For each action, we should:

- identify if it has variable size or fixed size (or close enough to fixed size, e.g., a `uint64` will have variable-length encoding but it's close enough to fixed size for these purposes)
- if fixed size, pick a representative size and hardcode that in the gas schedule
- if variable size, the gas schedule should be `.encode_to_vec().len()`

I think we don't have any variable-size actions that need "building". For instance a `Spend` needs building, but we need to compute a gas cost from a `SpendPlan` before building it, but a spend is fixed-size so we can use a hardcoded value. Or an `IbcRelay` has variable length, but does not need building, the plan is the action itself, so there's no problem. We just want to avoid having to put `action.encode_to_vec()` on an action type where there's a plan.

**Comment by aubrika** [comment]
> For each action, we should:

> * identify if it has variable size or fixed size (or close enough to fixed size, e.g., a `uint64` will have variable-length encoding but it's close enough to fixed size for these purposes)

is this done by determining if there are any variable-size data types (e.g. Vec<>) or via some other means?

> * if fixed size, pick a representative size and hardcode that in the gas schedule

what does "representative size" mean here?

> I think we don't have any variable-size actions that need "building". 

Are those actions that need "building" the ones with bespoke `ActionPlan`s, or does that mean another thing?

> For instance a `Spend` needs building, but we need to compute a gas cost from a `SpendPlan` before building it, but a spend is fixed-size so we can use a hardcoded value. Or an `IbcRelay` has variable length, but does not need building, the plan is the action itself, so there's no problem. 

As above, I'm unsure how to interpret this section exactly - it seems like there's a matrix here of variable/fixed length vs. needs building/doesnt need building, but also possibly a third variable of "gas cost is determined from the ActionPlan" vs. "gas cost is determined from the Action itself" - so I guess I'm unclear if there are 4 distinct cases to handle here, or 8 distinct cases.

> We just want to avoid having to put `action.encode_to_vec()` on an action type where there's a plan.

And by "plan" here, does that mean `ActionPlan`? Because as I understand it every `Action` has an associated plan.

**Comment by aubrika** [comment]
closed by #4116
